src/hazardous/aead/xchacha20poly1305_stream.rs

```diff
@@ -233,6 +234,14 @@ impl SecretStreamXChaCha20Poly1305 {
 		dst_out: &mut [u8],
 		tag: Tag,
 	) -> Result<(), UnknownCryptoError> {
+		// The largest addition done on length
+		// of plaintext. Other non-checked additions
+		// should not be able to overflow if this
+		// does not panic.
+		assert!((CHACHA_BLOCKSIZE as u64)
+			.checked_add(plaintext.len() as u64)
+			.is_some());
```

Comment:
I don't think this asserts are even necessary.
If i understand it correctly chach20 panics if it encrypts more then 2^32-1 * 64 bytes.  This is less then the message length where a overflow in the addition (if done on u64) begins to happen.

Reply:
You're right, this is not the right approach. This makes no sense in `seal_chunk` because the addition we're checking for comes after the call to ChaCha20. However, the preceding additions, on 64-bit targets, would still be able to overflow it seems. Specifically L246 and L256. 

What if we instead in `seal_chunk` checked for `assert!(SECRETSTREAM_XCHACHA20POLY1305_ABYTES.checked_add(msglen).is_some())`, because this is the most added to `msglen` without casting to `u64`. If I'm not mistaken, this way we can be sure that the remaining additions done in `seal_chunk` and subsequently in `generate_auth_tag` will not overflow, both on 32 and 64-bit targets. What do you think?

Reply:
Do you prefer asserts or checked_add + unwrap for catching overflows ?
checked_add seems to generate a bit better assembly https://gcc.godbolt.org/z/Nec_kf

Reply:
I like this, it's cleaner than what my approach would've ended up with. Feel free to add these changes.